### PR TITLE
feat(modals): Keyboard focus capture for modals

### PR DESCRIPTION
**Pull request by karansapolia**
Implement keyboard focus capture for modals in
content-server. Whenever a modal is opened,
track tab and shift+tab key combinations to keep
keyboard focus within modals and to cycle through
focusable modal elements until esc is pressed
and modal closes.

fixes: #597
fixes: #636

continuation of https://github.com/mozilla/fxa-content-server/pull/7102

@shane-tomlinson @vbudhram @lmorchard, would like to have your thoughts on this.

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/views/mixins/modal-panel-mixin.js [hunk 1]
```diff
@@ -45,6 +45,44 @@ module.exports = {
 
     this._boundBlockerClick = this.onBlockerClick.bind(this);
     $('.blocker').on('click', this._boundBlockerClick);
+
+    /**
+     * Trap keyboard focus when modal opens.
+     */
+    const modal = this.$el[0];
+
+    modal.addEventListener('keydown', tabClick);
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/mixins/modal-panel-mixin.js]: Can you add an `events` property to the mixin and add the handler there? That will ensure event housekeeping is taken care of, e.g., listeners are removed when the view is destroyed. Doing so might require ignoring keydown events while the panel is closed.

Comment (karansapolia) [reply packages/fxa-content-server/app/scripts/views/mixins/modal-panel-mixin.js]: Alright, let me do that.

**Comment by karansapolia** [comment]
> It has some problems on the "Account picture" section, when the modal is opened no element is focused

That seems to be because there is a hidden ['input' element](https://github.com/mozilla/fxa/blob/b6e0c65b98d1d37471da9c71f3ab116a2258db03/packages/fxa-content-server/app/scripts/templates/settings/avatar_change.mustache#L14) which gets the focus when the modal opens. You can execute 'document.activeElement' in the console to check that. If you tab after the modal opens, the upload button should be the next active element now.

> and I'm only able to tab through the "Done" button and the hidden file input. I'm wondering why it won't let me tab to the Upload and Camera button, though once I upload an image, all the controls are tab-able.

Don't know why that is happening, I am unable to replicate that on my device. Let me check if there are platform differences or other reasons.

**Comment by shane-tomlinson** [comment]
@karansapolia are you still interested in getting the PR across the line? It's a great idea, it'd be a shame to let the code come to nothing. If you no longer want to work on this, either one of us can take it over or we can close this PR down for another day.

**Comment by karansapolia** [comment]
> Can you add an `events` property to the mixin and add the handler there? That will ensure event housekeeping is taken care of, e.g., listeners are removed when the view is destroyed. Doing so might require ignoring keydown events while the panel is closed.

Well actually I tried what you said, but after adding the handler to the events property somehow it stopped catching click events altogether. I thought it might be a silly error on my side and read through how Backbone works and tried to find ways online but could not understand why it is suddenly not working at all. I was too ashamed to ask, felt like a very small issue which I should have been able to solve myself. That's what kept the PR where it is now.

> @karansapolia are you still interested in getting the PR across the line? It's a great idea, it'd be a shame to let the code come to nothing. If you no longer want to work on this, either one of us can take it over or we can close this PR down for another day.

I really want to get this through. I would need a little help here. I tried what you said, will share the code in a while too. Adding the handler to the events property makes no listener fire up when the modal opens. Surely some silly understanding error on my side, but would like to discuss with you whenever you get the time.

**Comment by shane-tomlinson** [comment]
Superceded by #2911. Your approach was used by @LZoog, thank you very much for the contribution here @karansapolia!

**Pull request closed without merging**